src/javascript/crypto/e2e/vrf/vrf.js

```diff
@@ -0,0 +1,200 @@
+      pk.length !== e2e.vrf.PUBLIC_KEY_SIZE_) {
+    return false;
+  }
+
+  var ed25519 = e2e.vrf.ed25519_;
+
+  // copy(vrf[:], vrfBytes)
+  // copy(pk[:], pkBytes)
+  // copy(c[:32], proof[:32])
+  // copy(t[:32], proof[32:64])
+  // copy(iiB[:], proof[64:96])
+  var c = new e2e.BigNum(proof.slice(0, 32).reverse());
+  var t = new e2e.BigNum(proof.slice(32, 64).reverse());
+  var iiB = proof.slice(64, 96);
+
+  // hash := sha3.NewShake256()
+  // hash.Write(iiB[:]) // const length
+  // hash.Write(m)
+  // var hCheck [Size]byte
+  // hash.Read(hCheck[:])
+  // if !bytes.Equal(hCheck[:], vrf[:]) {
+  //  return false
+  // }
+  // hash.Reset()
+
+  var hash = e2e.vrf.sha3.shake256(32).update(iiB).update(m).digest();
+  if (!e2e.compareByteArray(hash, vrf)) {
+    return false;
+  }
+
+  try {
+    // var P, B, ii, iic edwards25519.ExtendedGroupElement
+    // var A, hmtP, iicP edwards25519.ProjectiveGroupElement
+
+    // if !P.FromBytesBaseGroup(&pk) {
+    //   return false
+    // }
+    // if !ii.FromBytesBaseGroup(&iiB) {
+    //   return false
+    // }
+    var P = e2e.vrf.extra25519.fromBytesBaseGroup(pk);
+    if (P === false) {
+      return P;
+    }
+    var ii = e2e.vrf.extra25519.fromBytesBaseGroup(iiB);
+    if (ii === false) {
+      return ii;
+    }
+
+    // edwards25519.GeDoubleScalarMultVartime(&A, &c, &P, &t)
+    // A.ToBytes(&ABytes)
+    var ABytes = P.multiply(c).add(ed25519.curve.B.multiply(t)).toByteArray();
+
+    // hm := hashToCurve(m)
+    // edwards25519.GeDoubleScalarMultVartime(&hmtP, &t, hm, &[32]byte{})
+    // edwards25519.GeDoubleScalarMultVartime(&iicP, &c, &ii, &[32]byte{})
+    var hm = e2e.vrf.hashToCurve(m);
+    var zeroB = ed25519.curve.B.multiply(e2e.BigNum.ZERO);
+    var hmtP = hm.multiply(t).add(zeroB);
+    var iicP = ii.multiply(c).add(zeroB);
+
+    // iicP.ToExtended(&iic)
+    // hmtP.ToExtended(&B)
+    // edwards25519.GeAdd(&B, &B, &iic)
+    // B.ToBytes(&BBytes)
+    var BBytes = iicP.add(hmtP).toByteArray();
+
+
+    // var cH [64]byte
+    // hash.Write(ABytes[:]) // const length
+    // hash.Write(BBytes[:]) // const length
+    // hash.Write(m)
+    // hash.Read(cH[:])
+
+    // the two ABytes and BBytes must be of length 32
+    // ref: e2e.ecc.point.Ed25519.prototype.toByteArray
+    hash = e2e.vrf.sha3.shake256(64)
+        .update(ABytes)
+        .update(BBytes)
+        .update(m)
+        .digest();
+
+    // edwards25519.ScReduce(&cRef, &cH)
+    var cRef = ed25519.n.residue(new e2e.BigNum(hash.reverse()));
+
+    // return cRef == c
+    return cRef.isEqual(c);
+
+  } catch (e) {
+    return false;
+  }
+};
+
+
+/**
+ * Converts an 32-byte hashed output to an Ed25519 point using Elligator
+ * This point is multiplied by 8.
+ *
+ * Refer to Section 5.2 in http://elligator.cr.yp.to/elligator-20130828.pdf
+ * @param {!e2e.ByteArray} m A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.hashToCurve = function(m) {
```

Comment:
Please document that this is not constant time and thus should not be used with secret inputs. reason: hashToCurve isn't.